packages/apollo-gateway/src/__tests__/gateway/lifecycle-hooks.test.ts

```diff
@@ -0,0 +1,151 @@
+import * as books from '../__fixtures__/schemas/books';
+import * as inventory from '../__fixtures__/schemas/inventory';
+import * as product from '../__fixtures__/schemas/product';
+import * as reviews from '../__fixtures__/schemas/reviews';
+
+const services = [product, reviews, inventory, accounts, books];
+const serviceDefinitions = services.map((s, i) => ({
+  name: s.name,
+  typeDefs: s.typeDefs,
+  url: `http://localhost:${i}`,
+}));
+
+describe('lifecycle hooks', () => {
+  it('uses updateServiceDefinitions override', async () => {
+    const experimental_updateServiceDefinitions = jest.fn(
+      async (config: GatewayConfig) => {
+        return [serviceDefinitions, true];
+      },
+    );
+
+    const gateway = new ApolloGateway({
+      serviceList: serviceDefinitions,
+      experimental_updateServiceDefinitions,
+      experimental_didUpdateComputedFederationConfig: jest.fn(),
+    });
+
+    await gateway.load();
+
+    expect(experimental_updateServiceDefinitions).toBeCalled();
+  });
+
+  it('calls experimental_didFailComposition with a bad config', async done => {
+    const update = jest.fn(async (config: GatewayConfig) => {
+      return [[serviceDefinitions[0]], true];
+    });
+
+    const experimental_didFailComposition = jest.fn();
+
+    const gateway = new ApolloGateway({
+      experimental_updateServiceDefinitions: update,
+      experimental_didFailComposition,
+    });
+
+    try {
+      await gateway.load();
+    } catch (e) {
+      const callbackArgs = experimental_didFailComposition.mock.calls[0][0];
+      expect(callbackArgs.serviceList).toHaveLength(1);
+      expect(callbackArgs.errors).toMatchInlineSnapshot(`
+                Array [
+                  [GraphQLError: [product] Book -> \`Book\` is an extension type, but \`Book\` is not defined in any service],
+                ]
+            `);
+
+      expect(experimental_didFailComposition).toBeCalled();
+      done();
+    }
+  });
+
+  fit('calls experimental_didUpdateComposition on schema update', async done => {
+    const update = jest.fn();
+    update.mockImplementationOnce(async (config: GatewayConfig) => {
+      return [serviceDefinitions, true];
+    });
+    update.mockImplementationOnce(async (config: GatewayConfig) => {
+      const services = serviceDefinitions.filter(s => s.name !== 'books');
+      // overwrite books service with a similar 'book' service
+      return [
+        [
+          ...services,
+          {
+            name: 'book',
+            typeDefs: books.typeDefs,
+            url: 'http://localhost:32542',
+          },
+        ],
+        true,
+      ];
+    });
+
+    const experimental_didUpdateComposition = jest.fn();
+
+    const gateway = new ApolloGateway({
+      experimental_updateServiceDefinitions: update,
+      experimental_pollInterval: 10,
+      experimental_didUpdateComposition,
+    });
+
+    const { interval } = await gateway.load();
+    await new Promise(resolve => setTimeout(resolve, 30));
```

Comment:
It's nice to use a mock timer here and fast-forward, but that might be a lot of tech debt to tackle here... not sure.

Reply:
nit: 25? This seems like a potentially quite flaky test

Reply:
Yep! Definitely. I was trying this for testing. I can't get the interval to run, was trying extending the wait to see if it fixed it :)